src/exit_handler/test/test_exit_handler_intel_x64_eapis.cpp

```diff
@@ -4704,3 +4726,189 @@ eapis_ut::test_handle_vmcall_json_wrmsr_wrmsr_access_log_denied()
+
+    vmcs->enable_cr3_load_hook(&(exit_handler_ut::generic_cr_access_hook));
+
+    ehlr = setup_ehlr(vmcs);
+
+    g_state_save.rax = 0xbabe1ade;
+
+    RUN_UNITTEST_WITH_MOCKS(mocks, [&]
+    {
+        this->expect_true(guest_cr3::get() == 0xbeefface);
+
+        this->expect_no_exception([&] { ehlr->dispatch(); });
+
+        this->expect_true(guest_cr3::get() == 0xfeedface);
+    });
+
+    /* Test mov from cr3 */
+    vmcs = setup_vmcs(
+               mocks,
+               exit_reason::basic_exit_reason::control_register_accesses,
+               19 // MOV rax, cr3
+           );
+
+    vmcs->enable_cr3_store_hook(&(exit_handler_ut::generic_cr_access_hook));
+
+    ehlr = setup_ehlr(vmcs);
+
+    g_state_save.rax = 0xbabe1ade;
+
+    RUN_UNITTEST_WITH_MOCKS(mocks, [&]
+    {
+        this->expect_no_exception([&] { guest_cr3::set(0xbabebabe); });
+
+        this->expect_no_exception([&] { ehlr->dispatch(); });
+
+        this->expect_true(g_state_save.rax == 0xfeed1337);
+    });
+
+
+    /* Test mov to cr4 */
+    vmcs = setup_vmcs(
+               mocks,
+               exit_reason::basic_exit_reason::control_register_accesses,
+               4 // MOV cr0, rax
+           );
+
+    vmcs->enable_cr4_load_hook(&(exit_handler_ut::generic_cr_access_hook), 0, 0);
+
+    ehlr = setup_ehlr(vmcs);
+
+    g_state_save.rax = 0xbabe1ade;
+
+    RUN_UNITTEST_WITH_MOCKS(mocks, [&]
+    {
+        this->expect_true(guest_cr4::get() == 0xbeefface);
+
+        this->expect_no_exception([&] { ehlr->dispatch(); });
+
+        this->expect_true(guest_cr4::get() == 0xfeedface);
+    });
+
+    /* Test mov to cr8 */
+
+    vmcs = setup_vmcs(
+               mocks,
+               exit_reason::basic_exit_reason::control_register_accesses,
+               8 // MOV cr8, rax
+           );
+
+    vmcs->enable_cr8_load_hook(&(exit_handler_ut::generic_cr_access_hook));
+
+    ehlr = setup_ehlr(vmcs);
+
+    g_state_save.rax = 0xbabe1ade;
+
+    RUN_UNITTEST_WITH_MOCKS(mocks, [&]
+    {
+        this->expect_no_exception([&] {intel_x64::cr8::set(0xbeefface); });
+        this->expect_true(intel_x64::cr8::get() == 0xbeefface);
+
+        this->expect_no_exception([&] { ehlr->dispatch(); });
+
+        this->expect_true(intel_x64::cr8::get() == 0xfeedface);
+    });
+
+    /* Test mov from cr8 */
+    vmcs = setup_vmcs(
+               mocks,
+               exit_reason::basic_exit_reason::control_register_accesses,
+               24 // MOV rax, cr8
+           );
+
+    vmcs->enable_cr8_store_hook(&(exit_handler_ut::generic_cr_access_hook));
+
+    ehlr = setup_ehlr(vmcs);
+
+    g_state_save.rax = 0xbabe1ade;
+
+    RUN_UNITTEST_WITH_MOCKS(mocks, [&]
+    {
+        this->expect_no_exception([&] { intel_x64::cr8::set(0xbabebabe); });
+
+        this->expect_no_exception([&] { ehlr->dispatch(); });
+
+        this->expect_true(g_state_save.rax == 0xfeed1337);
+    });
+
+}
+
+using namespace exit_qualification;
+using namespace control_register_access;
```

Comment:
Please remove these and fully qualify below. We have run into issues doing this too much